### PR TITLE
Enumerable Methods Project

### PR DESCRIPTION
**Pull request by ZeenatLawal**
In this project we:

- Implemented our own version of some Ruby enumerable methods
- Created the following methods:
#my_each
#my_each_with_index
#my_select
#my_all?
#my_any?
#my_none?
#my_count
#my_map
#my_inject
- Tested the #my_inject with a #multiply_els
- Modified #my_map method to take a proc instead
- Modified #my_map method to take either a proc or a block

### CODE REVIEW

**Review by umairarshadbutt — COMMENTED** [review]
# Status: INVALID :red_circle: 

**Hi @ZeenatLawal  and @Vagyasri**
You should make sure that your repo looks professional before review. Please take a step back and read a [lesson here](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/66243642/). Then apply 3 simple rules from this lesson (it will be easy as you can use a template provided there). Before the next Code Review Request update a descriptive README file.
This Code Review Request will be marked as invalid in your Dashboard, so please submit a new one once you are ready!
##

 **If you are stuck at any point or have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help**
HAPPY CODING :computer:
**[Umair Arshad (Code Reviewer)](https://umairarshad.me/)**
[Twitter](https://twitter.com/its_UmairArshad) | [LinkedIn](https://www.linkedin.com/in/umair-arshad-butt/) | [GitHub](https://github.com/umairarshadbutt)

File: enumerables.rb [hunk 1]
```diff
@@ -0,0 +1,126 @@
+# rubocop:disable Style/For, Style/GuardClause
+# rubocop:disable  Metrics/ModuleLength
+
+module Enumerable
+  def my_each
+    for index in self
+      yield index
+    end
+  end
```
Comment (Berabjesus) [inline enumerables.rb]: - [x]  returns an Enumerator if no block is given. 
- If there is no block given there should be a conditional statement that checks if there is not block is given. Look into how you can return the enum itself if no block is given.

**Comment by ZeenatLawal** [comment]
Corrections have been made @umairarshadbutt

File: enumerables.rb [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    else
+      my_each do |i|
+        if i
+          return true
+        else
+          return false
+        end
+      end
+    end
+  end
```
Comment (Berabjesus) [inline enumerables.rb]: - [x] Your `my_all?` is not accepting ranges.
```
  false_block = proc { |num| num > 9 };
  range = Range.new(1,9)
  p range.my_all?(&false_block) #error
  p range.all?(&false_block) #false
```

- [x] `my_all?` should return false when there is a false in the array.
```
  p [1, false, 'hi', []].my_all? # True
  p [1, false, 'hi', []].all? #false
```

- [x]  when a class is passed as an argument returns true if all of the collection is a member of such class::Integer
```
 [1,2,3,4].my_all?(Numeric) #returns error
```

- [x] Should accept and check for regex.
````
 p %w[dog door rod blade].my_all?(/d/) #error
 p %w[dog door rod blade].all?(/d/)  #true
````

- [x] It should check for a pattern shown below
```
 p [3,3,3,3,3,3].my_all?(3)  //error
 p [3,3,3,3,3,3].all?(3) //true
```

File: enumerables.rb [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+  def my_any?
+    if empty?
+      true
+    elsif include?(nil) || include?(false)
+      false
+    else my_select { |i| return true if yield i }
+    end
+  end
```
Comment (Berabjesus) [inline enumerables.rb]: - [x] Check if `my_any?` returns the right result.
```
  block = proc { |num| num > 9 }
  p [1,2,3,4,5].my_any?(&block) // [ ]
  p [1,2,3,4,5].any?(&block) // false
```

File: enumerables.rb [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+    if num.instance_of?(Integer) || num.instance_of?(String)
+      result = num
+      my_each { |item| result = yield(result, item) }
+    elsif sym.instance_of?(Symbol)
+      result = num
+      my_each { |item| result = result.send(sym, item) }
+    end
+    result
+  end
+end
```
Comment (Berabjesus) [inline enumerables.rb]: - [x]  searches for the longest word in an array of strings

```
  search = proc { |memo, word| memo.length > word.length ? memo : word }
  p %w[dog door rod blade].my_inject(&search) # Error
  p %w[dog door rod blade].inject(&search) // "blade"
```

- [x]  when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range

```
  range = Range.new(5, 50) 
  p range.my_inject { |prod, n| prod * n } # 0
  p range.inject { |prod, n| prod * n } # 1267253883404724085150525340252698701849068398706688000000000000
```

- [x]  when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method

```
  p [1,2,3].my_inject(:+) # nil
  p [1,2,3].inject(:+) # 6
```

**Review by Berabjesus — CHANGES_REQUESTED** [review]
## Status: CHANGES REQUIRED :warning: :construction: 
Hi @ZeenatLawal  and @Vagyasri great start to this project. Very nicely put PR title and description.

#### However, since this is a very important section in ruby, there is a detailed test for it and you do have a few changes to carry out before you proceed so please take a look at the things you have to work on and their respective review comments below.

### General
- [x] Whenever appropriate, If there is no block given, there should be a conditional statement that checks for that. 

### my_each
- [x]  returns an Enumerator if no block is given

### my_each_with_index
- [x]  returns an Enumerator if no block is given

### my_select
- [x]  returns an Enumerator if no block is given

### my_all?
- [x]  returns false if the block returns false or nil::range
- [x] returns true when none of the collection members are false or nil.
- [x]  when a class is passed as an argument returns true if all of the collection is a member of such class::Integer. Make sure your enum method accepts class.
- [x] when a Regex is passed as an argument returns true if all of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
- [x] when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern

### my_any? (Take the same explanation from `my_all?` below)
- [x] returns false if the block ever returns a value false or nil
- [x]  returns false if the block ever returns a value false or nil::range
- [x]  when no block or argument is given returns true if at least one of the collection is not false or nil
- [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class::Numeric
- [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class::Integer
- [x]  when a Regex is passed as an argument returns true if any of the collection matches the Regex
- [x]  when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern
- [x]  when a pattern other than Regex or a Class is given returns true if any of the collection matches the pattern

### my_none? (Take the same explanation from `my_all?` below)
- [x]  when no block or argument is given returns false only if one of the collection members is true
- [x]  when a class is passed as an argument returns true if none of the collection is a member of such class
- [x]  when a class is passed as an argument returns true if none of the collection is a member of such class::Numeric
- [x]  when a class is passed as an argument returns false if any of the collection is a member of such class
- [x]  when a Regex is passed as an argument returns true if none of the collection matches the Regex
- [x]  when a Regex is passed as an argument returns false if any of the collection matches the Regex
- [x]  when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern
- [x] when a pattern other than Regex or a Class is given returns false only if one of the collection matches the pattern

### my_map 
- [x] returns an Enumerator if no block is given

### my_inject
- [x]  searches for the longest word in an array of strings
- [x]  when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range
- [x]  when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
- [x]  when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
- [x]  when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range

#### Work on the above errors and request for another CR.

##**If you have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help** 
**HAPPY CODING :computer:[Bereket Beshane](https://bereketababu.me)** | [LinkedIn](https://linkedin.com/in/bereketbeshane) | [GitHub](https://github.com/Berabjesus)

File: enumerables.rb [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+      my_each { |i| return false if i == false || i.nil? }
+    elsif args.instance_of?(Class)
+      my_each { |i| return false if i.class.superclass != args && i.class != args }
+    elsif args.instance_of?(Regexp)
+      my_each { |i| return false unless args.match(i) }
+    else
+      my_each { |i| return false if i != args }
+    end
+    true
+  end
```
Comment (Berabjesus) [inline enumerables.rb]: - [x] returns true if the block never returns false or nil
- [x]  returns false if the block returns false or nil
```
p [1,2,3,4,5].my_all?(proc { |num| num <= 9 }) #  flase
p [1,2,3,4,5].all?(proc { |num| num <= 9 }) # true

  block = proc { |num| num <= 9 }
  p [1,2,3,4,5].my_all?(&block) # error
  p [1,2,3,4,5].all?(&block) # true
```

- [x] returns false if the block returns false or nil::range
```
 range = Range.new(1,5)
  block = proc { |num| num > 9 }
  range.my_all?(&block) # error
  range.all?(&block) # false
```

- [x] does not mutate the original array
```
  arr = [1,2,3,4,5]
  arr.my_all? { |num| num + 1 } # error
```

File: enumerables.rb [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+      my_each { |i| return true if i == true || !i.nil? }
+    elsif args.instance_of?(Class)
+      my_each { |i| return true if i.instance_of?(args) || i.class.superclass == args }
+    elsif args.instance_of?(Regexp)
+      my_each { |i| return true if args.match(i) }
+    else
+      my_each { |i| return true if i == args }
+    end
+    false
+  end
```
Comment (Berabjesus) [inline enumerables.rb]: - [x] when no block or argument is given returns false if at least one of the collection is not true

```
  p [nil, false, nil, false].any? # false
  p [nil, false, nil, false].my_any? # true
```

File: enumerables.rb [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+    elsif num.instance_of?(Symbol) || num.instance_of?(String)
+      result = sym
+      my_each do |item|
+        result = result.nil? ? item : result.send(num, item)
+      end
+      result
+    end
+    result
+  end
+end
```
Comment (Berabjesus) [inline enumerables.rb]: - [x] raises a "LocalJumpError" when no block or argument is given
```
[1,2,3,4].inject # LocalJumpError
```

**Review by Berabjesus — CHANGES_REQUESTED** [review]
## Status: REQUIRED CHANGES :warning: :construction: 
Hi @ZeenatLawal and @Vagyasri it's great to see you have fixed most errors from the previous review. 

### However, there are still some changes that you need to make before this project is deemed complete. I believe, while you were working on the previous review, you might have introduced a few new bugs that were not there initially. Please have a look at the issues below.
  
### my_all? 

- [x]  returns true if the block never returns false or nil
- [x]  returns false if the block returns false or nil
- [x] returns false if the block returns false or nil::range
- [x] does not mutate the original array

### my_any? 
- [x] when no block or argument is given returns false if at least one of the collection is not true

### my_inject 
- [x] raises a "LocalJumpError" when no block or argument is given

#### **If you have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help** 
**HAPPY CODING :computer:[Bereket Beshane](https://bereketababu.me)** | [LinkedIn](https://linkedin.com/in/bereketbeshane) | [GitHub](https://github.com/Berabjesus)

**Comment by ZeenatLawal** [comment]
All corrections have been made @Berabjesus

**Review by jubaan — CHANGES_REQUESTED** [review]
[![Microverse][microverse-shield]][microverse-url]
<div>
<h1 align="center"> ⚠️ 🛠️ ⚠️ 🛠️ Changes Required 🛠️ ⚠️ 🛠️ ⚠️ </h1>
<br />
<p><b>"Fail fast, fail often"</b> and don't let this discourage you, because <b>only by failing do we learn to succeed</b>.</p>
<details open="open">
  <summary>Review Summary</summary>
  <ol>
    <li><a href="#congrats"> 💯 Awesome work 👏🏼 👏🏼 </a></li>
    <li><a href="#mandatory"> ⚠️ Things to Improve </a> 🚨 🚨 </li>
    <li><a href="#optional"> 💡Nice to have and other comments ✍🏼 ✍🏼 </a></li>
    <li><a href="#contact"> 🧔🏼 About your Reviewer </a> (you don't know it... but we are just automated bots 🤖) </li>
    <li><a href="#help"> 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼 </a></li>
  </ol>
</details>
</div>

<h3 id="congrats"> 1. :100: Awesome work</h3>
<details close="close">
<summary> [click to open] Big or little achievements, but achievements at the end. <b>Feel proud that you're learning and improving everyday.</b></summary>



</details>

<hr>
<h3 id="mandatory"> 2. ⚠️ Things to Improve (or the To-Do list to get the project approved) </h3> 
<details open="open">
<summary> [click to open] Working on these observations will help you advance and get closer to the goal 🎯, postpone them or do them half-ways and they'll haunt you 👻 👻... FOREVER!! (It will feel like that, trust me). </summary>


- [ ] `my_all?`, `my_inject`, and `multiply_els` don't work as expected.

- [ ] `#my_all?` when no block or argument is given returns true when none of the collection members are false or nil
- [ ] `#my_all?` when a class is passed as an argument returns false if any of the collection is not a member of such class
- [ ] `#my_all?` when a Regex is passed as an argument returns false if any of the collection does not match the Regex
- [ ] `#my_all?` when a pattern other than Regex or a Class is given returns false if any of the collection does not match the pattern
- [ ] `#my_inject` searches for the longest word in an array of strings
- [ ] `#my_inject` does not mutate the original array
- [ ] `#my_inject` when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block
- [ ] `#my_inject` when a block is given without an initial value combines all elements of enum by applying a binary operation, specified by a block::range
- [ ] `#my_inject` when a block is given with an initial value combines all elements of enum by applying a binary operation, specified by a block
- [ ] `#my_inject` when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method
- [ ] `#my_inject` when a symbol is specified without an initial value combines each element of the collection by applying the symbol as a named method:range
- [ ] `#my_inject` when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method
- [ ] `#my_inject` when a symbol is specified with an initial value combines each element of the collection by applying the symbol as a named method::range
- [ ] `#multiply_els` accepts an array as an argument and multiplies all the elements of the array together using #my_inject



</details>
<p><b>NOTE</b>This list is or could be divided in what is show in this comment and inline comments posted below.<p>
<hr>

<h3 id="optional"> 3. 💡 Nice to have and other comments </h3>
<details close="close">
<summary> [click to open] </h3></summary>


- [x] Nothing more to add.


</details>

<h3 id="contact"> 4. 🧔🏼 About your Reviewer </h3>
<details close="close">
<summary> [click to open] Say "hello world!" on Slack: @jubaan</summary>
<br />
<div align="center">
  <a src="https://www.github.com/jubaan">
    Julio Añoveros
  </a>
  <p>
    Julio Añoveros | @jubaan</br></br>
    <b>Full-stack Web Developer</b> and</br>
    🤖 Code Reviewer at <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a>:rocket:
  </p>
</div>
</details>

<h3 id="help"> 5. 🆘 <b> 😭 Feeling stuck? or Need help</b>? 🙏🏼 🙏🏼</h3>
<details close="close">
  <summary> [click to open] <b>Feel stuck?</b> </summary>
  <br>
  <p> First breath... just do it. Take 15-20 min away from the computer, stretch, meditate, do some exercise, drink water, or do whatever is necessary to distract your mind from the obstacle.</p>
  <p><b>Sometimes you just need to get away to see things with a new perspective.</b></p>
</details>

<details close="close">
  <summary> [click to open] <b>Need help?</b> </summary>
  <br>
  <p> Send me a DM in <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> Students Slack, you'll find me as @jubaan. If I'm not available there are other Code Reviewers you can contact. They'll help as well.</p>
  <p>Remember that <a href="https://www.microverse.org/?grsf=kfm2ej">Microverse</a> has other channels where you can get an answer to your questions:</p>
  <ol>
    <li>The <b>students channel 🧑‍🎓 </b> in slack,</li>
    <li>your <b>teammates 👥 </b>,</li>
    <li>your <b>mentor 👨🏼 👩🏼 </b>, and</li>
    <li><b>Student Success 🏫 </b>.</li>
  </ol>
  <p><b> DON'T hesitate to ask! </b></p>
</details>
<p>If all of the above doesn't help... get a <a href="https://rubberduckdebugging.com/">rubber duck</a> ... it would be great to own one 😄. </p>

<h2 align="center"> Keep up the great work :muscle: and Happy Coding :smiley: :computer:!! </h1>

[microverse-shield]: https://img.shields.io/badge/-microverse-%236F23FF?style=for-the-badge
[microverse-url]: https://www.microverse.org/?grsf=kfm2ej
[bootstrap-shield]: https://img.shields.io/badge/Bootstrap-white?style=for-the-badge&logo=bootstrap
[bootstrap-url]: https://www.microverse.org/?grsf=kfm2ej

**Review by Berabjesus — APPROVED** [review]
# Status: PROJECT APPROVED ✔️ 
Congratulations @ZeenatLawal and @Vagyasri you have implelemented all the required changes. Excellent job on this project Keep it up.  :champagne: :champagne:

**All the required tasks have been fulfilled in this project. I will go ahead and mark this as complete. 👍** 


I wish you all the very best in your next tasks:1st_place_medal:
#### **If you have any questions, let me know by leaving a comment here and mentioning me. I will be happy to help** 
**HAPPY CODING :computer:[Bereket Beshane](https://bereketababu.me)** | [LinkedIn](https://linkedin.com/in/bereketbeshane) | [GitHub](https://github.com/Berabjesus)

**Pull request merged**